File: demos_python/12_color_tool_refactor/command_loop.py
from color_tool_app import ColorToolApp


class CommandLoop:
    def __init__(self, app: ColorToolApp) -> None:
        self.app = app

    @classmethod
    def command_exit(cls) -> None:
        print("Exiting...")

    @classmethod
    def command_unknown(cls, command: str) -> None:
        print(f"Unknown command: {command}")

    def run(self) -> None:
        while True:
            command = input("Please enter a command > ")

            match command:
                case "add":
                    self.app.add_color()
                case "show":
                    self.app.show_colors()
                case "exit":
                    CommandLoop.command_exit()
                    break
                case _:
                    CommandLoop.command_unknown(command)

File: demos_rust/12_color_tool_refactor/src/command_loop.rs
use crate::color_tool_app::ColorToolApp;
use std::io::{stdin, stdout, Write};

// 'a is a lifetime parameter that is used to ensure that the reference to the ColorToolApp
// instance is valid for the duration of the CommandLoop instance.
pub struct CommandLoop<'a> {
    app: &'a mut ColorToolApp,
}

// 'a is a lifetime parameter that is used to ensure that the reference to the ColorToolApp
// instance is valid for the duration of the CommandLoop instance.
impl<'a> CommandLoop<'a> {
    pub fn new(app: &'a mut ColorToolApp) -> CommandLoop {
        CommandLoop { app }
    }

    fn get_command() -> String {
        let mut command = String::new();

        print!("Enter command: ");
        stdout().flush().unwrap();
        stdin().read_line(&mut command).unwrap();

        command.trim().to_string()
    }

    fn command_exit() {
        println!("Exiting...");
    }

    fn command_unknown(command: &str) {
        println!("Unknown command: {}", command);
    }

    pub fn run(&mut self) {
        loop {
            // Self refers to the CommandLoop struct itself.
            // Self::get_command is like a class method in Python
            let command = Self::get_command();
            match command.as_str() {
                "add" => {
                    self.app.add_color();
                }
                "show" => {
                    self.app.show_colors();
                }
                "exit" => {
                    Self::command_exit();
                    break;
                }
                _ => {
                    Self::command_unknown(command.as_str());
                }
            }
        }
    }
}
